asyncpg/connection.py

```diff
@@ -1356,15 +1359,27 @@ def _drop_global_statement_cache(self):
             # and https://github.com/MagicStack/asyncpg/issues/76
             # for discussion.
             #
+            await self._maybe_close_bad_connection()
             self._drop_global_statement_cache()
             if self._protocol.is_in_transaction() or not retry:
                 raise
             else:
                 return await self._do_execute(
                     query, executor, timeout, retry=False)
+        except:
```

Comment:
As far as I see from my (admittedly cursory) glance of the custom exceptions here, it looks like they all inherit from `Exception`. At the very least, inheriting from that would be great, although it might be even better to go through and mark some as "potentially signifying a bad connection" and catching only those.

Just grabbing a random exception from that file as an example: `DeprecatedFeature` probably shouldn't lead to killing&restarting the connection, though (as you commented) `ReadOnlySQLTransactionError` definitely could indicate this case and I am +1 on reacting to that case.